NOISSUE - Add report documentation (#52)
absmach/magistrala-docs: docs/dev-guide/reports.md

---
title: Reports
---

The **Reports Service** in Magistrala enables automated generation and scheduling of data reports from connected devices and sensors. This service allows users to collect, aggregate, and distribute metrics in PDF and CSV formats through email or direct download.

## Architecture
The Reports Service operates through three main components:

1. **Report Configurations**: Define what data to collect and how to process it
2. **Scheduler**: Handles recurring report generation based on defined schedules

> 3. **Generator Engine**: Creates human-readable reports in multiple formats

Reviewer: ```suggestion
3. **Generator Engine**: Generates human-readable reports in various formats.
```

| To            | string    | End time (relative/absolute)             | ✅       |
| FileFormat    | Format    | Output format (PDF/CSV)                  | Optional |
| Aggregation   | AggConfig | Data processing configuration            | Optional |

type AggConfig struct {
    AggType string  // "SUM", "AVG", "MIN", "MAX", "COUNT"
    Interval string
}
```

| Field     | Type        | Description                              | Required              |
|-----------|-------------|------------------------------------------|-----------------------|
| AggType   | Aggregation | Processing type (MAX/MIN/SUM/AVG/COUNT)  | If aggregation needed |
| Interval  | string      | Time window (e.g., "1h", "5m")           | If AggType specified  |

### Example configurations:

- Daily sales report at 8 AM: **DAILY** + **08:00**
- Weekly energy summary every Monday: **WEEKLY** + **00:00**
- Monthly inventory report: **MONTHLY** + **09:00**

### Report Generation

Data Collection
1. Connects to Magistrala's time-series database
2. Collects data using configured:
    - Time range (From/To)
    - Aggregation method
    - Metric filters

> 3. Supports complex queries across multiple devices/channels

Reviewer: ```suggestion
3. Supports complex queries across multiple devices and channels
```

| Best For	        | Printing/sharing	                | Programmatic analysis              |

Example Data Representation

PDF Table:

| Time                |  Value | Unit |    Protocol    | Subtopic    |
|---------------------|--------|------|----------------|-------------|
| 2024-03-15 09:30:00 |  23.4  | °C   |      http      |  room1      |
| 2024-03-15 09:35:00 |  45.2  | °C   |      http      |  room1      |

Equivalent CSV:

```csv
Timestamp,Value,Unit,Protocol,Subtopic
2024-03-15T09:30:00Z,23.4,°C,http,room1
2024-03-15T09:35:00Z,45.2,°C,http,room1
```

### Email Integration

```go
type EmailSetting struct {
    To      []string  // Recipient addresses
    Subject string    // Email subject line
}
```

| Field     | Type     | Description                              | Required |
|-----------|----------|------------------------------------------|----------|
| To        | []string | Recipient email addresses                | ✅       |
| Subject   | string   | Email subject line                       | Optional |
| Content   | string   | Email body content                       | Optional |

> **NOTE:**

> > Automatically sends generated reports as email with body text summarizing report contents.

Reviewer: ```suggestion
> Automatically sends generated reports via email, including a summary of the report contents in the body.
```

### API Operations

Base URL: `http://localhost:9008/{domainID}/reports`

1. Create Report Configuration

Endpoint: `POST /configs`

```bash
curl --location http://localhost:9008/domains/{domainID}/reports/configs \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer $ACCESSTOKEN' \
--data '{
    "name": "lab 1 report",

>     "descripion": "lab 1 sensors report",

Reviewer: ```suggestion
    "description": "lab 1 sensors report",
```

    "config": {
        "from": "now()-5d",
        "to": "now()",
        "aggregation": {
            "agg_type":"MAX",
            "interval":"1s"
        }
    },
    "metrics": [
        {
            "channel_id": "{{CHANNELID}}",
            "client_id": "{{THINGID}}",
            "name": "current"
            "subtopic": "lab/room1"
        }
    ]
    "schedule": {
        "start_datetime": "2025-04-28T00:00:00.000Z",
        "time": "0001-01-01T20:30:00.000Z",
        "recurring": "daily",
        "recurring_period": 1
    },
    "email": {
        "to": ["team@example.com"],

>         "subject": "Weeky Lab Report"

Reviewer: ```suggestion
        "subject": "Weekly Lab Report"
```

    "domain_id": "88c2bc9a-ce3b-4dfc-804d-219177cb9a75",
    "schedule": {
        "start_datetime": "2025-04-28T00:00:00Z",
        "time": "0001-01-01T20:30:00Z",
        "recurring": "daily",
        "recurring_period": 1
    },
    "config": {
        "from": "now()-5d",
        "to": "now()",
        "aggregation": {
            "agg_type": "max",
            "interval": "1s"
        }
    },
    "email": {
        "to": [
            "team@example.com"
        ],
        "subject": "Weekly Lab Report"
    },
    "metrics": [
        {
            "channel_id": "f0e052ce-ef01-49b9-862c-2ecbc911f0a1",
            "client_id": "69a9b488-523a-4805-aae9-123febaf83f5",
            "name": "current",
            "format": ""
        }
    ],
    "status": "enabled",
    "created_at": "2025-04-28T12:59:13.768526Z",
    "created_by": "14c2a388-310d-402b-ad8e-8c0b7d9f81f1",
    "updated_at": "0001-01-01T00:00:00Z"
}
```

2. Generate Report
Endpoint: `POST /`

```bash
curl -X POST http://localhost:9008/domains/{domainID}/reports \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer $ACCESSTOKEN' \
--data '{
    "name": "lab 1 report",

>     "descripion": "lab 1 sensors report",

Reviewer: ```suggestion
    "description": "lab 1 sensors report",
```

    "config": {
        "from": "now()-5d",
        "to": "now()",
        "aggregation": {
            "agg_type":"MAX",
            "interval":"1s"
        }
    },
    "metrics": [
        {
            "channel_id": "{{CHANNELID}}",
            "client_id": "{{THINGID}}",
            "name": "current"
            "subtopic": "lab/room1"
        }
    ]
}'
```

> Expected reponse:

Reviewer: ```suggestion
Expected response:
```

                "name": "current",
                "format": ""
            },
            "messages": [
                {
                    "subtopic": "lab/room1",
                    "protocol": "http",
                    "name": "current",
                    "unit": "A",
                    "time": 1276020072001000000,
                    "value": 1.33
                },
                {
                    "subtopic": "lab/room1",
                    "protocol": "http",
                    "name": "current",
                    "unit": "A",
                    "time": 1276020071001000000,
                    "value": 1.5
                }
            ]
        }
    ]
}
```

> **Note**: This API can be used to configure multiple actions and download reports.
> The available actions are `view`, `download` and `email`. The default action is `view`. To change the action you can pass the action parameter:

```bash
curl --location 'http://localhost:9008/bd1bb2c5-ce78-4456-8725-bd1beab80250/reports?action=download' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer $ACCESSTOKEN' \
--data '{
    "name": "lab 1 report",

>     "descripion": "lab 1 sensors report",

Reviewer: ```suggestion
    "description": "lab 1 sensors report",
```

    "config": {
        "from": "now()-5d",
        "to": "now()",
        "file_format": "pdf"
    },
    "metrics": [
        {
            "channel_id": "a246bea7-dc05-49f1-bb61-c08c44a7df33",
            "client_id": "e2ab8673-8599-435f-8d42-62903d279492",
            "name": "current"       
        }
    ]  
}'
```

> Expected reponse:

Reviewer: ```suggestion
Expected response:
```

![report_pdf](../diagrams/pdf.png)

> **Note**: The supported formats are `csv` and `pdf` which are set in the config field.

To send the generated report via email it can be done as follows:

```bash
curl --location 'http://localhost:9008/bd1bb2c5-ce78-4456-8725-bd1beab80250/reports?action=email' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer $ACCESSTOKEN' \
--data-raw '{
    "name": "lab 1 report",

>     "descripion": "lab 1 sensors report",

Reviewer: ```suggestion
    "description": "lab 1 sensors report",
```

---
title: Reports
---

> The **Reports Service** in Magistrala enables automated generation and scheduling of data reports from connected devices and sensors. This service allows users to collect, aggregate, and distribute metrics in PDF and CSV formats through email or direct download.

Reviewer: ```suggestion
The **Reports Service** in Magistrala provides a streamlined way to generate and schedule data reports from connected devices and sensors. It allows users to collect, aggregate, and export metrics in PDF and CSV formats, either via email or direct download.
```
